net: tcp2: Lock connection when running from work queue

jukkar opened the pull request:
We run various TCP function from work queue. Make sure the
connection lock is taken before accessing the connection.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

Possible fix to #28587

jukkar committed:
net: tcp2: Lock connection when running from work queue

We run various TCP function from work queue. Make sure the
connection lock is taken before accessing the connection.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ethernet: Check that header is valid

Saw this crash with heavily loaded system in mimxrt1050_evk:

<err> os: ***** MPU FAULT *****
<err> os:   Data Access Violation
<err> os:   MMFAR Address: 0xc
<err> os: r0/a1:  0x80000ab0  r1/a2:  0x800f6a60  r2/a3:  0x00000000
<err> os: r3/a4:  0x800f72a0 r12/ip:  0x00000000 r14/lr:  0x6000eb43
<err> os:  xpsr:  0x41000000
<err> os: Faulting instruction address (r15/pc): 0x6000dc82
<err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
<err> os: Current thread: 0x80001a18 (rx_workq)
<err> os: Halting system

Where the fault at 0x6000dc82 points to ethernet_recv()

	uint16_t type = ntohs(hdr->type);
6000dc82:	89ab      	ldrh	r3, [r5, #12]

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: conn: Check that TCP pointer is valid

Saw this crash with heavily loaded system in nucleo_f767zi:

<err> os: ***** MPU FAULT *****
<err> os:   Data Access Violation
<err> os:   MMFAR Address: 0x0
<err> os: r0/a1:  0x800f6d30  r1/a2:  0x80005d84  r2/a3:  0x00000006
<err> os: r3/a4:  0x00000000 r12/ip:  0x00000001 r14/lr:  0x60013f69
<err> os:  xpsr:  0x61000000
<err> os: Faulting instruction address (r15/pc): 0x60014304
<err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
<err> os: Current thread: 0x80001a18 (rx_workq)
<err> os: Halting system

Where the fault at 0x60014304 points to net_conn_input()

   } else if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
	src_port = proto_hdr->tcp->src_port;
60014300:	f8d9 3000 	ldr.w	r3, [r9]
60014304:	881a      	ldrh	r2, [r3, #0]

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Saw some crashes with heavily loaded system, added new commits fixing those.

jukkar commented:
Removed the crash checks as they cannot happen any more after commit c54a511d26d0127f60ef6f2fbee6b3757defe627 is merged.

mniestroj commented on subsys/net/ip/tcp2.c:
```diff
@@ -413,6 +417,11 @@ static void tcp_send_process(struct k_work *work)
 	}
+
+out:
+	if (conn) {
```
I don't like these `if (conn)` checks in this function. The only way they will be true is when `conn = NULL;` is set in this function. I think that it is simpler to just exit early instead of setting `conn = NULL;` and then checking `if (conn)` twice.

jukkar replied:
I suppose we could return in line 394 instead of setting `conn=NULL`. I think the idea here has been to have only one exit point in this function so we have the out: label at the end. I would leave this as is.

pfalcon replied:
> I think the idea here has been to have only one exit point in this function so we have the out: label at the end.

+1. This is common "exception handling" pattern in C (including pattern of label naming). Wouldn't make sense to (mis)"optimize" it for this particular case, risking that it may be problematic in case of further changes.

mniestroj replied:
So maybe in case of `conn = NULL;` lets jump after all `if (conn)` statements, as there is no reason to go through them? Such goto would be of course equivalent of `return;`, but technically there would be still one exit point.

Another idea would be to add `bool unref = false;`, which would be set `unref = true` instead of `conn = NULL`. Something like:
```
static void tcp_send_process(struct k_work *work)
{
	struct tcp *conn = CONTAINER_OF(work, struct tcp, send_timer);
	struct net_pkt *pkt;
	bool unref = false;

	k_mutex_lock(&conn->lock, K_FOREVER);

	...

	if (conn->in_retransmission) {
		if (conn->send_retries > 0) {
			...
		} else {
			unref = true;
			goto out;
		}
	} else {
		...
	}

	if (conn->in_retransmission) {
		k_delayed_work_submit(&conn->send_timer, K_MSEC(tcp_rto));
	}

out:
	k_mutex_unlock(&conn->lock);

	if (unref) {
		tcp_conn_unref(conn);
	}
}
```

Regarding label naming https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions documents how good label names should look like. Such naming helps if there are more labels, as each one describes what it does. `out` isn't particularly good name to say "unlock packet and return".

pfalcon replied:
> Regarding label naming https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions documents how good label names should look like. Such naming helps if there are more labels,

Right, it helps in case there're many labels, and in case there's one, it helps if it's named consistently, showing it's general purpose. And this purpose is effectively like "finally" clause in languages which support explicit exception handling. I won't argue here that "finally" would make a good label name, but if we have a pattern of using "out" for this purpose, sticking to it makes sense IMHO.

jukkar replied:
> Another idea would be to add `bool unref = false;`, which would be set `unref = true` instead of `conn = NULL`.

Sure, this looks good. I will send a new version.

mniestroj commented on subsys/net/ip/tcp2.c:
```diff
@@ -413,6 +417,11 @@ static void tcp_send_process(struct k_work *work)
 		k_delayed_work_submit(&conn->send_timer, K_MSEC(tcp_rto));
 	}
+
+out:
```
Maybe `unlock` will be a more meaningful name?

jukkar replied:
We have used the out: label in other functions so I would call this same here.

mniestroj commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -174,10 +174,16 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 	struct net_linkaddr *lladdr;
 	sa_family_t family;
 
+	if (hdr == NULL) {
```
`hdr` comes from `net_pkt_data(pkt)`. Are we sure that `pkt->buffer` is non-NULL? Maybe its safer to use `net_pkt_is_empty()`, that you introduced lately?

While at it I also wonder if we should check for `pkt->buffer->len >= hdr_len`.

jukkar replied:
> Are we sure that `pkt->buffer` is non-NULL?

That pointer is/has been checked from day one in the network stack, I am pretty sure it is not NULL at this point. The problem was the pkt->buffer->data pointer that could have been NULL if we run out of bufs.
I think it is more intuitive to check the `hdr` here instead of net_pkt as we are later using the hdr variable.

> While at it I also wonder if we should check for `pkt->buffer->len >= hdr_len`.

Yes, it looks like we never check that there is enough bytes for Ethernet header. I will add this check.

mniestroj replied:
>The problem was the pkt->buffer->data pointer that could have been NULL if we run out of bufs.

If we run out of bufs (meaning net_buf structures), then `pkt->frags` (AKA `pkt->buffer`) is going to be NULL. I don't see how `pkt->buffer->data` can be NULL for net_buf_fixed type of net_bufs. What am I missing?

jukkar replied:
> If we run out of bufs (meaning net_buf structures), then `pkt->frags` (AKA `pkt->buffer`) is going to be NULL. I don't see how `pkt->buffer->data` can be NULL for net_buf_fixed type of net_bufs. What am I missing?

I have been looking the tcp2 and net_buf code quite a lot lately, and it is a bit difficult to see all the interactions when the out-of-buffer situation happens. One thing I noticed that looks a bit suspicious, is that the net_buf refcounting is not atomic which could cause some issues or not.

jukkar committed:
net: ethernet: Make sure Ethernet header is in the recv pkt

Check that Ethernet header is in the first net_buf fragment.
This is very unlikely to happen as device driver is expected
to only deliver proper Ethernet frames to upper stack.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
